File: app/src/main/kotlin/com/foodshare/features/notifications/domain/repository/NotificationRepository.kt
package com.foodshare.features.notifications.domain.repository

import com.foodshare.features.notifications.domain.model.PaginatedNotificationsResult
import com.foodshare.features.notifications.domain.model.UserNotification
import kotlinx.coroutines.flow.Flow

/**
 * Repository interface for notification data operations.
 *
 * SYNC: Mirrors Swift NotificationRepository
 */
interface NotificationRepository {

    /**
     * Fetch notifications with pagination.
     */
    suspend fun getNotifications(
        userId: String,
        limit: Int = 20,
        offset: Int = 0
    ): Result<List<UserNotification>>

    /**
     * Fetch paginated notifications with metadata.
     */
    suspend fun getPaginatedNotifications(
        userId: String,
        limit: Int = 20,
        offset: Int = 0
    ): Result<PaginatedNotificationsResult>

    /**
     * Get unread notification count.
     */
    suspend fun getUnreadCount(userId: String): Result<Int>

    /**
     * Mark a notification as read.
     */
    suspend fun markAsRead(notificationId: String): Result<Unit>

    /**
     * Mark all notifications as read.
     */
    suspend fun markAllAsRead(userId: String): Result<Unit>

    /**
     * Delete a notification.
     */
    suspend fun deleteNotification(notificationId: String): Result<Unit>

    /**
     * Observe new notifications in real-time.
     */
    fun observeNotifications(userId: String): Flow<UserNotification>
}

File: Sources/FoodShare/Features/Notifications/Domain/Repositories/NotificationRepository.swift
//
//  NotificationRepository.swift
//  Foodshare
//
//  Protocol for notification data access
//

import Foundation

// MARK: - Paginated Notifications Result

struct PaginatedNotificationsResult: Sendable {
    let notifications: [UserNotification]
    let unreadCount: Int
    let totalCount: Int
    let hasMore: Bool

    static let empty = PaginatedNotificationsResult(
        notifications: [],
        unreadCount: 0,
        totalCount: 0,
        hasMore: false,
    )
}

// MARK: - Notification Repository Protocol

protocol NotificationRepository: Sendable {
    /// Fetch all notifications for a user
    func fetchNotifications(for userId: UUID, limit: Int, offset: Int) async throws -> [UserNotification]

    /// Fetch unread notifications count
    func fetchUnreadCount(for userId: UUID) async throws -> Int

    /// Fetch paginated notifications with unread count in a single call (server-side)
    func fetchPaginatedNotifications(
        for userId: UUID,
        limit: Int,
        offset: Int,
    ) async throws -> PaginatedNotificationsResult

    /// Mark a notification as read
    func markAsRead(notificationId: UUID) async throws

    /// Mark all notifications as read
    func markAllAsRead(for userId: UUID) async throws

    /// Delete a notification
    func deleteNotification(notificationId: UUID) async throws

    /// Subscribe to real-time notification updates
    func subscribeToNotifications(
        for userId: UUID,
        onNotification: @escaping @Sendable (UserNotification) -> Void,
    ) async
}
